pkg/content/content_clean.go

```diff
@@ -0,0 +1,398 @@
+package content
+
+import (
+	"bufio"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const (
+	JcrContentFile    = ".content.xml"
+	JcrMixinTypesProp = "jcr:mixinTypes"
+	JcrRootPrefix     = "<jcr:root"
+	PropPattern       = "^\\s*([^ =]+)=\"([^\"]+)\"(.*)$"
+	NamespacePattern  = "^\\w+:(\\w+)=\"[^\"]+\"$"
+	JcrRoot           = "jcr_root"
+)
+
+type Cleaner struct {
+	config *Opts
+}
+
+func NewCleaner(config *Opts) *Cleaner {
+	return &Cleaner{
+		config: config,
+	}
+}
+
+func (c Cleaner) prepare(root string) error {
+	if c.config.ParentsBackupEnabled {
+		return c.doParentsBackup(root)
+	}
+	return nil
+}
+
+func (c Cleaner) BeforeClean(root string) error {
+	if c.config.ParentsBackupEnabled {
+		return c.doRootBackup(root)
+	}
+	return nil
+}
+
+func (c Cleaner) Clean(root string) error {
+	err := c.flattenFiles(root)
+	if err == nil {
+		if c.config.ParentsBackupEnabled {
+			err = c.undoParentsBackup(root)
+		} else {
+			err = c.cleanParents(root)
+		}
+	}
+	if err == nil {
+		err = c.cleanDotContents(root)
+	}
+	if err == nil {
+		err = c.deleteFiles(root)
+	}
+	if err == nil {
+		err = c.deleteBackupFiles(root)
+	}
+	if err == nil {
+		err = deleteEmptyDirs(root)
+	}
+	return err
+}
+
+func eachFilesInDir(root string, processFileFunc func(path string) error) error {
+	infos, err := os.ReadDir(root)
+	for i := 0; i < len(infos) && err == nil; i++ {
+		if !infos[i].IsDir() {
+			err = processFileFunc(filepath.ToSlash(filepath.Join(root, infos[i].Name())))
```

Comment:
this is wrong; early return; each err in iteration in your code could be overridden by another one which does not make sense